File: lib/ai-reco/pricing.ts
/**
 * Tabela de preços manual dos modelos OpenRouter usados (consultada 2026-05-24).
 * Atualizar quando OpenRouter mudar (raro). Modelo ausente → custo = 0
 * (acceptable degradation: tracking continua, só custo_usd vira null/0).
 */
export const MODEL_PRICING_USD_PER_1M_TOKENS = {
  "deepseek/deepseek-r1": { in: 0.55, out: 2.19 },
  "deepseek/deepseek-v3.2": { in: 0.27, out: 1.10 },
  "anthropic/claude-sonnet-4.5": { in: 3.00, out: 15.00 },
} as const;

export function computeCostUsd(
  model: string,
  promptTokens: number,
  completionTokens: number,
): number {
  const p =
    MODEL_PRICING_USD_PER_1M_TOKENS[
      model as keyof typeof MODEL_PRICING_USD_PER_1M_TOKENS
    ];
  if (!p) return 0;
  return (promptTokens * p.in + completionTokens * p.out) / 1_000_000;
}

File: scripts/scraper/lib/scraper/ai_reco/pricing.rb
module AdamStats
  module Scraper
    module AiReco
      # Tabela de preços dos modelos OpenRouter usados (consultada 2026-05-24).
      # Atualizar quando OpenRouter mudar (raro). Modelo ausente → custo = 0
      # (degradação aceitável: tracking continua, só cost_usd vira 0).
      #
      # Espelha lib/ai-reco/pricing.ts — qualquer mudança deve ser sincronizada.
      module Pricing
        MODEL_PRICING_USD_PER_1M_TOKENS = {
          'deepseek/deepseek-r1'        => { in: 0.55, out: 2.19 },
          'deepseek/deepseek-v3.2'      => { in: 0.27, out: 1.10 },
          'anthropic/claude-sonnet-4.5' => { in: 3.00, out: 15.00 }
        }.freeze

        module_function

        # @param model [String] ex: 'deepseek/deepseek-r1'
        # @param prompt_tokens [Integer]
        # @param completion_tokens [Integer]
        # @return [Float] custo em USD (0 se modelo desconhecido)
        def compute_cost_usd(model, prompt_tokens, completion_tokens)
          p = MODEL_PRICING_USD_PER_1M_TOKENS[model]
          return 0 unless p

          (prompt_tokens.to_f * p[:in] + completion_tokens.to_f * p[:out]) / 1_000_000.0
        end
      end
    end
  end
end
